qutip/qip/circuit_latex.py

```diff
@@ -42,31 +47,121 @@
+_pdfcrop = _find_system_command(['pdfcrop'])
+
+
+def _force_remove(*filenames):
+    """rm -f`: try to remove a file, ignoring errors if it doesn't exist."""
+    for filename in filenames:
+        try:
+            os.remove(filename)
+        except FileNotFoundError:
+            pass
+
+
+_run_command = functools.partial(subprocess.run, check=True,
+                                 stdout=subprocess.DEVNULL,
+                                 stderr=subprocess.DEVNULL)
+_run_command.__doc__ = \
+    """
+    Run a command with stdout and stderr explicitly thrown away, raising
+    `subprocess.CalledProcessError` if the command returned a non-zero exit
+    code.
+    """
+
+if not _pdfcrop:
+    def _crop_pdf(_):
+        warnings.warn("Could not locate system 'pdfcrop':"
+                      " image output may have additional margins.")
+else:
+    def _crop_pdf(filename):
+        temporary = ".tmp." + filename
+        _run_command((_pdfcrop, filename, temporary))
+        os.rename(temporary, filename)
+
+
+def _convert_pdf(file_stem):
     """
-    Requires: pdflatex, pdfcrop, pdf2svg, imagemagick (convert)
+    'Convert' to pdf: since LaTeX outputs a PDF file, there's nothing to do.
     """
-    os.system("rm -f %s.tex %s.pdf %s.png" % (filename, filename, filename))
+    with open(file_stem + ".pdf", "rb") as file:
+        return file.read()
+
 
-    with open(filename + ".tex", "w") as file:
-        file.write(_latex_template % (_qcircuit_latex_min, code))
+# Record type to hold definitions of possible conversions - this is just for
+# reading convenience.
+_PossibleConverter = collections.namedtuple(
+    '_PossibleConverter',
+    ['file_type', 'dependency', 'executables', 'arguments', 'binary'],
+)
+CONVERTERS = {"pdf": _convert_pdf}
+_POSSIBLE_CONVERTERS = [
+    _PossibleConverter('png', 'ImageMagick', ['magick', 'convert'],
+                       arguments=('-density', '100'), binary=True),
+    _PossibleConverter('svg', 'pdf2svg', ['pdf2svg'],
+                       arguments=(), binary=False),
+]
 
-    os.system("pdflatex -interaction batchmode %s.tex" % filename)
-    os.system("rm -f %s.aux %s.log" % (filename, filename))
-    os.system("pdfcrop %s.pdf %s-tmp.pdf" % (filename, filename))
-    os.system("mv %s-tmp.pdf %s.pdf" % (filename, filename))
 
-    if format == 'png':
-        os.system("convert -density %s %s.pdf %s.png" % (100, filename,
-                                                         filename))
-        with open("%s.png" % filename, "rb") as f:
-            result = f.read()
+def _make_converter(possible):
+    """
+    Create the actual conversion function of signature
+        file_stem: str -> 'T,
+    where 'T is data in the format to be converted to.
+    """
+    which = _find_system_command(possible.executables)
+    if which is None:
+        return None
+    mode = "rb" if possible.binary else "r"
+
+    def converter(file_stem):
+        in_file = file_stem + ".pdf"
```

Comment:
Could you add NumPy-style documentation for `converter`? 
```
"""
It takes the stem string of the filename, transforms the corresponding pdf to png/svg and returns the plot.

Parameters
------------ 
...
Returns
--------
...
"""
```
 like other functions in QuTiP.

Reply:
Sure.  Despite not being marked as such, the whole module is pretty internal and the original `_latex_compile` didn't have a docstring, but I'll add one.